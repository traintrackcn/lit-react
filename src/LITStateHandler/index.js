import _StateHandler from './_state/Handler';
import LITBaseHandler from '../LITBaseHandler';

export default class LITStateHandler extends LITBaseHandler{
    
    constructor(path, ctx){
        super(path, ctx);
        this.setupPath();
    }

    setupPath() {
        try{
            this.p = this.getPath();
        }catch(e){
            
        }
    }

    getStateHandler() {
        if (!this._state) {
            const path = this.getPath();
            this._state = new _StateHandler(path);
            this.push(this._state);
        }
        return this._state;
    }

    getChildren() {
        if (!this._children) {
            this._children = [];
        }
        return this._children;
    }

    push(childH){
        const children = this.getChildren();
        children.push(childH);
        if (this.s) childH.s = this.s;
    }

    set s(value){
        this._s = value;
        const children = this.getChildren();
        children.forEach(child => {
            child.s = value;
        });
    }

    get s(){
        return this.getStateStore();
    }

    get(){
        const h = this.getStateHandler();
        return h.get();
    }

    set(value){
        const h = this.getStateHandler();
        h.set(value);
    }

    del(){
        const h = this.getStateHandler();
        h.del();
    }

    getValue(){
        const h = this.getStateHandler();
        return h.get_value();
    }

    setValue(value) {
        const h = this.getStateHandler();
        h.set_value(value);
    }

    delValue() {
        const h = this.getStateHandler();
        h.del_value();
    }

    getKey(key){
        const h = this.getStateHandler();
        const state = h.get();
        return state.get(key);
    }

    setKey(key, value){
        const h = this.getStateHandler();
        var state = h.get();
        state = state.set(key, value);
        this.set(state);
    }

    delKey(key){
        const h = this.getStateHandler();
        var state = h.get();
        state = state.delete(key);
        this.set(state);
    }

    getCollection(){
        const h = this.getStateHandler();
        return h.get_collection();
    }

    setCollection(value){
        const h = this.getStateHandler();
        h.set_collection(value);
    }

    select(index){
        const h = this.getStateHandler();
        h.select(index);
    }

    getSelectedIndex(){
        const h = this.getStateHandler();
        return h.get_selected_index();
    }






















    /**
     * @deprecated
     */
    get_value(){
        return this.getValue();
    }

    /**
     * @deprecated
     */
    set_value(value){
        this.setValue(value);
    }

    /**
     * @deprecated
     */
    del_value(){
        this.delValue();
    }

    /**
     * @deprecated
     */
    get_collection(){
        return this.getCollection();
    }

    /**
     * @deprecated
     */
    set_collection(value){
        this.setCollection(value);
    }

    /**
     * @deprecated
     */
    get_selected_index(){
        return this._state.get_selected_index();
    }

    /**
     * @deprecated
     */
    get_custom_value(key){
        const h = this.getStateHandler();
        const state = h.get();
        return state.get(key);
    }

    /**
     * @deprecated
     */
    set_custom_value(key, value){
        const h = this.getStateHandler();
        var state = h.get();
        state = state.set(key, value);
        this.set(state);
    }

    /**
     * @deprecated
     */
    del_custom_value(key){
        const h = this.getStateHandler();
        var state = h.get();
        state = state.delete(key);
        this.set(state);
    }

}