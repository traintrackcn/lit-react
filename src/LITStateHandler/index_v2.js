import _StateHandler from './_state/Handler';
import LITBaseHandler from '../LITBaseHandler';

export default class LITStateHandlerV2 extends LITBaseHandler{
    
    constructor(path, parent){
        super(path, parent);
    }

    get(){
        const h = this.getStateHandler();
        return h.get();
    }

    set(value){
        const h = this.getStateHandler();
        h.set(value);
    }

    del(){
        const h = this.getStateHandler();
        h.del();
    }

    getValue(){
        const h = this.getStateHandler();
        return h.get_value();
    }

    setValue(value) {
        const h = this.getStateHandler();
        h.set_value(value);
    }

    delValue() {
        const h = this.getStateHandler();
        h.del_value();
    }


    checkKey(key) {
        if (!key) throw new Error(`Cannot find key`);
        if (Array.isArray(key)) throw new Error(`Key should be string, current -> ${key}`);
    }

    getKey(key){
        try {
            this.checkKey(key);
            const h = this.getStateHandler();
            const state = h.get();
            // console.log('getKey state ->', key, state);
            return state.get(key);
        }catch(e) {
            throw e;
        }
    }

    setKey(key, value){
        try {
            this.checkKey(key);
            const h = this.getStateHandler();
            var state = h.get();
            // console.log('setKey state ->', key, value,state);
            state = state.set(key, value);
            this.set(state);
        }catch(e) {
            throw e;
        }
    }

    delKey(key){
        try {
            this.checkKey(key);
            const h = this.getStateHandler();
            var state = h.get();
            state = state.delete(key);
            this.set(state);
        }catch(e) {
            throw e;
        }
    }

    getCollection(){
        const h = this.getStateHandler();
        return h.get_collection();
    }

    setCollection(value){
        const h = this.getStateHandler();
        h.set_collection(value);
    }

    select(index){
        const h = this.getStateHandler();
        h.select(index);
    }

    getSelectedIndex(){
        const h = this.getStateHandler();
        return h.get_selected_index();
    }

    getStateHandler() {
        if (!this._state) {
            const path = this.getPath();
            this._state = new _StateHandler(path);
            this._state.s = this.getStateStore();
        }
        return this._state;
    }

}