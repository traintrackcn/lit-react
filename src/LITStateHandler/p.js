import LITPath from "../LITPath";

//for test purpose

export class RootPath extends LITPath{
    constructor(path){
        super(path);
        this.t = this.path.concat(['t']);
        this.a = this.path.concat(['a']);
    }
}



export default new RootPath(['r']);