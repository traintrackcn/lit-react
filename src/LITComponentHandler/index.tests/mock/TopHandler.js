import LITComponentHandler from "../..";
import p from '../../../__mock__/p';
import StateHandler from './StateHandler';
import ComponentHandler from './ComponentHandler';
import CommonHandler from './CommonHandler';


export default class TopHandler extends LITComponentHandler {


    getChildStateHandlerWithArrayPath() {
        if (!this._state) {
            this._state = new StateHandler(p.t, this);
        }
        return this._state;
    }

    getChildStateHandlerWithObjectPath() {
        if(!this._state1) {
            this._state1 = new StateHandler(p, this);
        }
        return this._state1
    }

    

    getChildComponentHandler() {
        if (!this._component) {
            this._component = new ComponentHandler(this);
        }
        return this._component;
    }

    getChildHandler() {
        if (!this._common) {
            this._common = new CommonHandler(this);
        }
        return this._common;
    }

    getCustomHandler() {
        return 'ok';
    }

}