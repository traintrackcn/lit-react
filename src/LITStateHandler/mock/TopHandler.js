import p from '../../__mock__/p';
import s from '../../s';
import LITStateHandlerV2 from '../index_v2';

export default class TopHandler extends LITStateHandlerV2{

    constructor(path, ctx, handlerClass) {
        super(path, ctx);

        this.setHandlerClass(handlerClass);
    }

    setHandlerClass(value) {
        this._handlerCls = value;
    }
    
    getHandlerClass() {
        return this._handlerCls;
    }

    getStateStore() {
        return s;
    }

    getH() {
        if (!this._h) {
            const Class = this.getHandlerClass();
            this._h = new Class(p.t, this);
        }
        return this._h;
    }
}