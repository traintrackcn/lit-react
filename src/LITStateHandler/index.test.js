
import TopHandler from './mock/TopHandler';
import LITStateHandlerV2 from './index_v2';
import LITStateHandler from '.';

describe('test state handler', () => {

    const clsArr = [LITStateHandlerV2, LITStateHandler];
    for (let index = 0; index < clsArr.length; index++) {
        const cls = clsArr[index];
        const topH = new TopHandler(null, null, cls);
        const h = topH.getH();
        const hName = h.getClassName();


        it(`test ${hName} setCollection`, () => {

            h.del();
    
            const collection = [
                'a', 'b', 'c'
            ];
            h.setCollection(collection);
            expect(h.get()).toMatchSnapshot();
            h.select(1);
            expect(h.getValue()).toMatchSnapshot();
    
            expect(h.get()).toMatchSnapshot();
        });
    
        it(`test ${hName} getPath`, () => {
            const expected = '["r","t"]';
            expect(JSON.stringify(h.getPath())).toBe(expected);
        })
    
        it(`test ${hName} set_value`, () => {
    
            h.del();
    
            const value = { 'key': 'value' }
            h.setValue(value);
            expect(h.get()).toMatchSnapshot();
            expect(h.getValue()).toMatchSnapshot();
        });

        
    }


});



