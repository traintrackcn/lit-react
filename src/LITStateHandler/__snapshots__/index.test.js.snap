// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test state handler test s ops 1`] = `"a"`;

exports[`test state handler test set_collection 1`] = `
Immutable.Map {
  "collection": Immutable.List [
    "a",
    "b",
    "c",
  ],
}
`;

exports[`test state handler test set_collection 2`] = `"b"`;

exports[`test state handler test set_collection 3`] = `
Immutable.Map {
  "collection": Immutable.List [
    "a",
    "b",
    "c",
  ],
  "index": 1,
  "value": "b",
}
`;

exports[`test state handler test set_value 1`] = `
Immutable.Map {
  "value": Immutable.Map {
    "key": "value",
  },
}
`;

exports[`test state handler test set_value 2`] = `
Immutable.Map {
  "key": "value",
}
`;
